### PR TITLE
Big rework

### PR DESCRIPTION
**Pull request by Vince-Chenal**
Various stuffs like:
- add tflint
- define TfChecks structs and interface
- factorize different event types via CheckEvent and GenericGithubEvent interface

### CODE REVIEW

File: pkg/github/check_api.go [hunk 1]
```diff
@@ -3,68 +3,73 @@ package github
...
-	var conclusion string
-	if status {
-		conclusion = "success"
-	} else {
-		conclusion = "failure"
+func (e *CheckEvent) CreateCheckRun(check terraform.TfCheck) (GhCheckRun, error) {
+	checkRunName := fmt.Sprintf("terraform-check %v %v", check.Name(), check.RelDir())
+	log.Info().Msgf("Create check run %s on repo %s PR %s", checkRunName, e.GetRepo().GetFullName(), e.GetPRUrl())
+
+	externalId := fmt.Sprintf("%s:%s", check.RelDir(), check.Name())
```
Comment (jb-abbadie) [inline pkg/github/check_api.go:19]: should probably be moved to a function `encodeExternalId` and `decodeExternalId` in the future
so we can keep encode/decode close to one another
could be nice to also start with a version like `1:<dir>:<check_name>` that way we can have `2:<dir>:<check_name>:<additional_field>`

File: pkg/github/check_api.go [hunk 2]
```diff
@@ -3,68 +3,73 @@ package github
...
+	actions := []*github.CheckRunAction{}
+	if conclusion != githubv4.CheckConclusionStateSuccess {
+		actions = check.FixActions()
 	}
-	actions := []*github.CheckRunAction{&fmtAction}
-	_, _, err := t.GhClient.Checks.UpdateCheckRun(context.TODO(),
-		t.Repo.Owner,
-		t.Repo.Name,
+
+	log.Info().Msgf("Update check run %s on repo %s PR %s", cr.Name, e.GetRepo().GetFullName(), e.GetPRUrl())
```
Comment (jb-abbadie) [inline pkg/github/check_api.go:59]: log levels will probably need to be reviewed in the future

File: pkg/github/structs.go [hunk 3]
```diff
@@ -15,9 +40,196 @@ type GhCheckRun struct {
...
-	Token      string
-	HeadBranch string
-	GhClient   *github.Client
+	GenericGithubEvent
+	repo     Repo
+	sha      string
+	token    string
+	branch   string
+	prUrl    string
+	ghClient *github.Client
```
Comment (jb-abbadie) [inline pkg/github/structs.go:49]: we could probably include a context in this struct and pass it to the rest of the checks

File: pkg/terraform/exec.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	ok, files, err := tf.FormatCheck(context.Background(), &tfexec.RecursiveOption{})
+	if err != nil {
+		log.Error().Err(err).Msg("error running terraform fmt check")
+		return false, ""
+	}
+	if !ok {
+		return false, "Your terraform formatting is wrong for the following files:\n" + strings.Join(
+			files,
+			"\n",
+		) + "\nplease run `terraform fmt` in the right dir or launch the `Trigger tf fmt` action ⬆️⬆️⬆️" + "\n\n" + "more info [here](https://www.terraform.io/docs/cli/commands/fmt.html)"
```
Comment (jb-abbadie) [inline pkg/terraform/exec.go:96]:
```suggestion
		) + "\nplease run `terraform fmt -recursive` in the right dir or launch the `Trigger tf fmt` action ⬆️⬆️⬆️" + "\n\n" + "more info [here](https://www.terraform.io/docs/cli/commands/fmt.html)"
```

File: pkg/terraform/exec.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+			errorString += "`" + diag.Range.Filename + "` line:`" + strconv.Itoa(diag.Range.Start.Line) + "`"
+			errorString += "\n"
+			errorString += diag.Summary
+			errorString += "\n"
+			errorString += diag.Detail
+			errorString += "\n"
+			errorString += "\n"
+		}
+		errorString += "\n"
+		return false, "The command `terraform validate` is failing on your repo with the following errors\n" + errorString
```
Comment (jb-abbadie) [inline pkg/terraform/exec.go:81]: maybe add something on how to reproduce locally with `terraform init -backend=false;terraform validate`

File: pkg/terraform/exec.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	if validationError.ErrorCount != 0 {
+		errorString := ""
+		for _, diag := range validationError.Diagnostics {
+			errorString += "`" + diag.Range.Filename + "` line:`" + strconv.Itoa(diag.Range.Start.Line) + "`"
+			errorString += "\n"
+			errorString += diag.Summary
+			errorString += "\n"
+			errorString += diag.Detail
+			errorString += "\n"
+			errorString += "\n"
```
Comment (jb-abbadie) [inline pkg/terraform/exec.go:78]: we could create a package to format stuff in Markdown with common function like this

File: pkg/github/check_api.go (hunk 1 above)
Comment (Vince-Chenal) [reply pkg/github/check_api.go:19]: done for the encode/decode functions.
we'll see later for improvements

File: pkg/github/check_api.go (hunk 2 above)
Comment (Vince-Chenal) [reply pkg/github/check_api.go:59]: Yep, not sure which ones should stay as Info, maybe the createCheckRun one is enough in Info

File: pkg/github/structs.go (hunk 3 above)
Comment (Vince-Chenal) [reply pkg/github/structs.go:49]: why not, let's discuss it IRL

File: pkg/terraform/exec.go (hunk 5 above)
Comment (Vince-Chenal) [reply pkg/terraform/exec.go:81]: ok will do

**Review by jb-abbadie — APPROVED** [review]

**Pull request merged**